Reorganize README & CONTRIBUTING and add testing docs (#350)
commercetools/ui-kit: CONTRIBUTING.md

## Development

### Testing

We split our testing in three parts: functionality, visual and the final bundle. We use separate tools for these, and we try to avoid overlap.

#### Functionality testing

- write tests from a user perspective, not from a dev perspective
- we test using `react-testing-library`
  - we have `test-utils` as a thin wrapper which sets up the context of tested components
  - we avoid `beforeEach` as it leads to test coupling
  - an `it` can describe the flow of a user
  - it is okay to have many `expect` statements in an `it`
  - code comments can help explain the `expect`s better
- avoid testing for visuals (classnames or the DOM structure)
  - the user does not care about the DOM structure or the classnames themselves
  - use visual testing instead (see below)
- if mocking is required, try to do it at the level the user interacts with the component
  - Examples
    - good: mocking network responses
    - bad: mocking a fetching component
  - Try to keep mocking minimal. The less we mock, and the later we mock, the more our tests will cover and the more confidence we gain that our components work

#### Visual testing

> - we will use a tool called [Percy](https://percy.io/) for visual regression testing

Reviewer: ```suggestion
- we use a tool called [Percy](https://percy.io/) for visual regression testing
```

- we test the components from the produced bundle
  - this gives us confidence that the bundling process works
  - this tests components under conditions closer to how consumers would use them
  - this helps us to avoid accidentally not exposing implemented components from the main `index.js` file
- we test components in multiple states in one snapshot
  - it is nice to see all states in one snapshot
  - it uses less snapshots (we have a monthly budget)
  - we use custom components called `Suite` and `Spec` to surround our component-under-test, which makes the review easier
  - we set a min-height for every `Spec` to ensure that changes in height of a component's state don't break the visual diff of states following below that

> - we can currently not "interact" with components (like opening a dropdown menu) before taking a screenshot. We might be able to solve this by using Puppeteer in the future, but we wanted to get up and running first.

Reviewer: ```suggestion
- at the moment we cannot "interact" with components (like opening a dropdown menu) before taking a screenshot. We might be able to solve this by using Puppeteer in the future, but we wanted to get up and running first.
```

#### Bundle testing

> We have test files ending with `*.bundlespec.js`. They are used to test the bundle produced by the rollup build. It is necessary to run the build before running these tests. We use these to further ensure that our bundling process works.

Reviewer: ```suggestion
We have test files ending with `*.bundlespec.js`. They are used to test the bundle produced by the rollup build, to ensure that our bundling process works.

Note that it is necessary to run the build before running these tests.
```

Author: This suggestion is missing the part which says that one needs to build before running the test. And I don't like the repeated use of "to ... to ...". So I left this one out. Thanks for the other two suggestions!

Reviewer: That line is there is the suggestion, didn’t you see it? 😅

- Documentation of prop-types and usage patterns
- Sensible defaults for props which are not a hard requirement

## Development

### Testing

We split our testing in three parts: functionality, visual and the final bundle. We use separate tools for these, and we try to avoid overlap.

#### Functionality testing

- write tests from a user perspective, not from a dev perspective
- we test using `react-testing-library`
  - we have `test-utils` as a thin wrapper which sets up the context of tested components
  - we avoid `beforeEach` as it leads to test coupling
  - an `it` can describe the flow of a user
  - it is okay to have many `expect` statements in an `it`
  - code comments can help explain the `expect`s better
- avoid testing for visuals (classnames or the DOM structure)
  - the user does not care about the DOM structure or the classnames themselves
  - use visual testing instead (see below)
- if mocking is required, try to do it at the level the user interacts with the component
  - Examples
    - good: mocking network responses

>     - bad: mocking a fetching component

Reviewer 2: maybe we can give a more relevant example to UI Kit, since it's unlikely we will ever have network requests in the lib itself 🤔

Author: I was trying to come up with a better example too, but I couldn't really. Maybe this is better?

Maybe:
- good: mocking `localStorage`
- bad: mocking a whole component which uses local storage

It's unlikely too. It's more about getting readers in the mindset that "the more tests reflect how a user uses an application, the more confidence they can give you" than about providing an actual example. So we should use mocks at the intersection with a user (DOM, network, local-storage, cookies) instead of mocking implementation parts of the app. Exceptions apply where it's must more convenient (like `ApolloMockProvider`) and worth the tradeoff.

>   - Try to keep mocking minimal. The less we mock, and the later we mock, the more our tests will cover and the more confidence we gain that our components work

Reviewer 2: I'm not sure what this means. Can you explain it in a different way? 😇

Author: I tried to reword it in https://github.com/commercetools/ui-kit/pull/350#discussion_r242853430. It boils down to the famous "[The more your tests resemble the way your software is used, the more confidence they can give you.](https://twitter.com/kentcdodds/status/977018512689455106)."

#### Visual testing

> 

Reviewer 2: maybe we should add the same as in Bundle Testing and explain that they need to run `build`?

- we test the components from the produced bundle
  - this gives us confidence that the bundling process works
  - this tests components under conditions closer to how consumers would use them
  - this helps us to avoid accidentally not exposing implemented components from the main `index.js` file
- we test components in multiple states in one snapshot
  - it is nice to see all states in one snapshot
  - it uses less snapshots (we have a monthly budget)
  - we use custom components called `Suite` and `Spec` to surround our component-under-test, which makes the review easier
  - we set a min-height for every `Spec` to ensure that changes in height of a component's state don't break the visual diff of states following below that
- at the moment we cannot "interact" with components (like opening a dropdown menu) before taking a screenshot. We might be able to solve this by using Puppeteer in the future, but we wanted to get up and running first.

#### Bundle testing

We have test files ending with `*.bundlespec.js`. They are used to test the bundle produced by the rollup build, to ensure that our bundling process works.

> Note that it is necessary to run the build before running these tests.

Reviewer 2: to run `build` before ...
